efficientdet/keras/train.py

```diff
@@ -0,0 +1,220 @@
+    else:
+      ds_strategy = tf.distribute.OneDeviceStrategy("device:CPU:0")
+
+  # Check data path
+  if FLAGS.mode in ('train',
+                    'train_and_eval') and FLAGS.training_file_pattern is None:
+    raise RuntimeError('You must specify --training_file_pattern for training.')
+  if FLAGS.mode in ('eval', 'train_and_eval'):
+    if FLAGS.validation_file_pattern is None:
+      raise RuntimeError('You must specify --validation_file_pattern '
+                         'for evaluation.')
+
+  params = dict(config.as_dict(),
+                model_name=FLAGS.model_name,
+                iterations_per_loop=FLAGS.iterations_per_loop,
+                model_dir=FLAGS.model_dir,
+                num_examples_per_epoch=FLAGS.num_examples_per_epoch,
+                strategy=FLAGS.strategy,
+                batch_size=FLAGS.train_batch_size,
+                eval_batch_size=FLAGS.eval_batch_size,
+                num_shards=FLAGS.num_cores,
+                val_json_file=FLAGS.val_json_file,
+                testdev_dir=FLAGS.testdev_dir,
+                mode=FLAGS.mode)
+
+  # set mixed precision policy by keras api which will generate mixed precision model
+  precision = utils.get_precision(params['strategy'], params['mixed_precision'])
+  policy = tf.keras.mixed_precision.experimental.Policy(precision)
+  tf.keras.mixed_precision.experimental.set_policy(policy)
+
+  def get_dataset(input_context, is_training, params):
+    global_batch_size = FLAGS.train_batch_size if is_training else FLAGS.eval_batch_size
+    batch_size = input_context.get_per_replica_batch_size(global_batch_size)
+    file_pattern = FLAGS.training_file_pattern if is_training else FLAGS.validation_file_pattern
+    params = copy.deepcopy(params)
+    update_params = dict(batch_size=batch_size) if is_training else dict(
+        eval_batch_size=batch_size)
+    params.update(update_params)
+    return dataloader.InputReader(
+        file_pattern,
+        is_training=is_training,
+        use_fake_data=FLAGS.use_fake_data,
+        max_instances_per_image=config.max_instances_per_image)(params)
+
+  train_dataset = ds_strategy.experimental_distribute_datasets_from_function(
+      lambda input_context: get_dataset(
+          input_context, is_training=True, params=params))
+  if FLAGS.validation_file_pattern:
+    eval_dataset = ds_strategy.experimental_distribute_datasets_from_function(
+        lambda input_context: get_dataset(
+            input_context, is_training=False, params=params))
+  else:
+    eval_dataset = None
+
+  with ds_strategy.scope():
+    model = EfficientDetNetTrain(params['model_name'], config)
+  ckpt_path = tf.train.latest_checkpoint(FLAGS.model_dir)
+  if ckpt_path:
+    model.load_weights(ckpt_path)
+  model.freeze_vars(params['var_freeze_expr'])
+  with ds_strategy.scope():
+    model.compile(
+        optimizer=get_optimizer(params),
+        box_loss=BoxLoss(params['delta'],
```

Comment:
I don't quite understand this. Should compile just take two parameters: optimizer and loss?  (the loss can be a dict or a string or a losses instance)?

Could you add the following test? (I tried it, but it failed)

  def test_train(self):
    config = hparams_config.get_detection_config('efficientdet-d0')
    config.batch_size = 1
    config.num_examples_per_epoch = 1
    x = np.random.random((1, 512, 512, 3))
    labels = {
        'box_targets_%d' % i: np.ones((1, 512 // 2**i, 512 // 2**i, 810))
        for i in range(3, 8)
    }
    labels.update({
        'cls_targets_%d' % i: np.random.random(
            (1, 512 // 2**i, 512 // 2**2, 36)) for i in range(3, 8)
    })
    labels.update({'mean_num_positives': 100})

    model = train_lib.EfficientDetNetTrain(config=config)
    params = config.as_dict()
    model.compile(
        optimizer=train_lib.get_optimizer(params),
    )
    outputs = model.train_on_batch(x, labels)

Reply:
I have overrided compile and implement train loop by myself, so you couldn't use train_on_batch and have to specific box_loss, box_iou_loss, class_loss

Reply:
I try to transfer to train_on_batch

Reply:
Aha, I see. What's the benefits to override compile (looks like it just simplifies the loss)? Should we combine those loss functions into a single loss, and then pass it to compile?

Reply:
I have migrated to train_on_batch